### DISCUSSION TITLE
Setting FontFamily at runtime (custom font)?

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#17199] [Q&A] [answered]
First of all, I want to say thank you to Avalonia for producing the following for packaging custom fonts:

https://docs.avaloniaui.net/docs/guides/styles-and-resources/how-to-use-fonts

It always been a confusing topic and error prone, so its incredibly helpful, however, I still I have no real idea how to set TextBlock.Font family at **runtime**. In the past, I have spent days pouring over this issue, and sometimes it works, and then it doesn't. I primarily want to do this at runtime, and all the examples seem to be XAML.

Please have patience with me on this.

So lets suppose I wish to use a custom font I've packaged as a resource, so I include the following the csproj:

    <ItemGroup>
        <AvaloniaResource Include="Assets\**" />
    </ItemGroup>

Then in the App.axaml:

    <Application.Resources>
        <FontFamily x:Key="NunitoFont">avares://GoogleFonts/Assets/Fonts#Nunito</FontFamily>
    </Application.Resources>

So what should FontFamily be at runtime?

TextBlock.FontFamily = "NunitoFont"  ?
TextBlock.FontFamily = "#NunitoFont"  ?

Or something else?

PS.

How is FontFamily handled if I do something like:

TextBlock.FontFamily = "serif"
TextBlock.FontFamily = "monospace"

This doesn't seem to do anything, so I'm guessing if falls to "Inter".

What happens if I set just:

TextBlock.FontFamily = "Inter" ?

Does it find "Inter", or does it not find it, but then falls back to Inter?

This is relevant where the Family needs to be set from a drop down box at the UI.

Thank you

**Comment by stevemonaco** [comment]
> So what should FontFamily be at runtime?

```cs
TextBlock.FontFamily = new FontFamily(@"avares://AvaloniaApplication/Assets/Fonts#Nunito"); // For a resource-based font
TextBlock.FontFamily = new FontFamily("Times New Roman"); // For a system font
```

I tend to store these in a `static class` instead. eg.

```cs
public static class AppResources
{
    public static FontFamily Nunito { get; } = new FontFamily(@"avares://AvaloniaApplication/Assets/Fonts#Nunito");
}
```

However, I'm not able to create+add an `EmbeddedFontCollection` and simply do `new FontFamily("Nunito")`. I think this is because the following code does only considers System fonts instead of looping all of its `_fontCollections`. The `FontFamily` key is null in my testing:

https://github.com/AvaloniaUI/Avalonia/blob/aabd038e88951a992124ddfcb2be06ee6daaf4a5/src/Avalonia.Base/Media/FontManager.cs#L130-L138

So I can't get something like this to work:

```cs
var fc = new EmbeddedFontCollection(new Uri("fonts:App", UriKind.Absolute), new Uri(@"avares://AvaloniaApplication/Assets", UriKind.Absolute));
FontManager.Current.AddFontCollection(fc);

text.FontFamily = new FontFamily("Nunito");
```

When I try using `FontManager.Current.TryGetGlyphTypeface`, I receive "Segoe UI" on Win11, so I guess a default system font:

```cs
var ff = new FontFamily(new Uri("fonts:App"), "Nunito");
var tf = new Typeface(ff);
FontManager.Current.TryGetGlyphTypeface(tf, out var gtf);
```

That's about as far as I can troubleshoot for now. (Actual testing was done with Anton SC as Nunito looks too similar to Inter)

**Reply by kuiperzone** [reply]
So if I do this:

'TextBlock.FontFamily = new FontFamily(@"avares://AvaloniaApplication/Assets/Fonts#Nunito"); // For a resource-based font'

Do I still need to put this in the App.axaml:

```
<Application.Resources>
    <FontFamily x:Key="NunitoFont">avares://GoogleFonts/Assets/Fonts#Nunito</FontFamily>
</Application.Resources>
```

?

I would like to thank you for your detailed answer. I think I need a static class to hold known font names, as you suggest. Cheers! Andy

**Reply by stevemonaco** [reply]
No, not necessary to put into resources if you always specify the full asset uri with `new FontFamily(...)`.

I prefer static classes for "resources" that are never swapped or need to be overridden lower in the visual tree. Less resource keys, compile-time errors instead of runtime errors, and it's easy to consume from both XAML (`{x:Static ...}`) and code.

**Reply by Gillibald** [reply]
If you define a font family without a source we implicitly map the name to the system font collection.

https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Media/FontManager.cs#L21

So `FontFamily="Segoe UI"` becomes `FontFamily="fonts:SystemFonts#Segoe UI"`

**Reply by kuiperzone** [reply]
Thank you.

What does the prefix "fonts:" do? As in `FontFamily="fonts:...`?

If I put some kind of resource in my App.Axaml, can I use `FontFamily="fonts:MyFonts#Nunito`?

**Reply by stevemonaco** [reply, marked as answer]
Ah, good to know. Thanks!

For anybody following along, this is possible today (fixing my example above):

```cs
var fc = new EmbeddedFontCollection(new Uri("fonts:App", UriKind.Absolute), new Uri(@"avares://AvaloniaApplication25/Assets", UriKind.Absolute));
FontManager.Current.AddFontCollection(fc);

text.FontFamily = new FontFamily("fonts:App#Nunito");
```

It leaks a bit of font storage knowledge vs simply "Nunito", so I do wish any added `FontCollection`s would be searched before the default font is used, but I might be bikeshedding.

I still prefer the `static class` approach if your entire set of custom fonts is known at compile-time and come from resources.

**Reply by kuiperzone** [reply]
Thank you @stevemonaco . Your answer really helped!
